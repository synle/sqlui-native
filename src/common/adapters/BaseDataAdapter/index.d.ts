import { SqluiCore } from 'typings';
export declare const MAX_CONNECTION_TIMEOUT = 3000;
export default abstract class BaseDataAdapter {
    protected connectionOption: string;
    dialect?: SqluiCore.Dialect;
    constructor(connectionOption: string);
    protected getConnectionString(): string;
    static getConnectionParameters(connection: string): import("connection-string-parser").IConnectionStringParameters;
    static resolveTypes(inputItem: any, incomingTypeConverter?: (type: string, value: any) => string): Record<string, SqluiCore.ColumnMetaData>;
    static inferTypesFromItems(items: any[]): SqluiCore.ColumnMetaData[];
    static inferSqlTypeFromItems(items: any[], toDialectHint?: string): SqluiCore.ColumnMetaData[];
}
