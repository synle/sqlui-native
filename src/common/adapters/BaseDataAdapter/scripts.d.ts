import IDataScript from 'src/common/adapters/IDataScript';
import { SqlAction, SqluiCore } from 'typings';
export declare function getDivider(): SqlAction.Output;
export default abstract class BaseDataScript implements IDataScript {
    dialects: string[];
    isDialectSupported(dialect?: string): boolean;
    getConnectionFormInputs(): string[][];
    getIsTableIdRequiredForQuery(): boolean;
    getSyntaxMode(): string;
    supportMigration(): boolean;
    supportCreateRecordForm(): boolean;
    supportEditRecordForm(): boolean;
    supportVisualization(): boolean;
    getDialectType(dialect?: SqluiCore.Dialect): SqluiCore.Dialect;
    getDialectName(dialect?: SqluiCore.Dialect): string;
    getDialectIcon(dialect?: SqluiCore.Dialect): string;
    getTableScripts(): SqlAction.TableActionScriptGenerator[];
    getDatabaseScripts(): SqlAction.DatabaseActionScriptGenerator[];
    getConnectionScripts(): SqlAction.ConnectionActionScriptGenerator[];
    getSampleConnectionString(dialect?: SqluiCore.Dialect): string;
    getSampleSelectQuery(actionInput: SqlAction.TableInput): SqlAction.Output | undefined;
    getCodeSnippet(connection: SqluiCore.ConnectionProps, query: SqluiCore.ConnectionQuery, language: SqluiCore.LanguageMode): string;
}
