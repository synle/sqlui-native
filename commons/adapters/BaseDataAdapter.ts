import { ConnectionStringParser } from 'connection-string-parser';
import { SqluiCore } from '../../typings';

export default abstract class BaseDataAdapter {
  protected connectionOption: string;
  public dialect?: SqluiCore.Dialect;

  constructor(connectionOption: string) {
    this.connectionOption = connectionOption as string;
  }

  /**
   * get dialect string from a connection string
   * @param {string} connection in the uri scheme (cassandra://localhost:9042)
   * @return {string} the dialect, in this case, it's cassandra
   */
  static getDialect(connection: string): string | undefined {
    try {
      return connection.substr(0, connection.indexOf(':')).toLowerCase();
    } catch (err) {
      return undefined;
    }
  }

  /**
   * this method will parse connection string into connection parameters
   *
   * @param {string} connection in the uri scheme (cassandra://localhost:9042)
   */
  static getConnectionParameters(connection: string) {
    const dialect = BaseDataAdapter.getDialect(connection);
    switch (BaseDataAdapter.getDialect(connection)) {
      case 'mysql':
      case 'mariadb':
      case 'mssql':
      case 'postgres':
      case 'sqlite':
      case 'cassandra':
      case 'mongodb':
        if (dialect) {
          const connectionStringParser = new ConnectionStringParser({
            scheme: dialect,
            hosts: [],
          });
          return connectionStringParser.parse(connection);
        }
        break;
    }

    // not supported
    return undefined;
  }
}
