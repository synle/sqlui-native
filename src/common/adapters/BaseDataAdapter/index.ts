import { ConnectionStringParser } from 'connection-string-parser';
import { getDialectType } from 'src/common/adapters/DataScriptFactory';
import { SqluiCore } from 'typings';

export const MAX_CONNECTION_TIMEOUT = 3000;

export default abstract class BaseDataAdapter {
  protected connectionOption: string;
  public dialect?: SqluiCore.Dialect;

  constructor(connectionOption: string) {
    this.connectionOption = connectionOption as string;
    this.dialect = getDialectType(connectionOption);
  }

  protected getConnectionString(): string {
    return this.connectionOption.replace(/^[a-z0-9+]+:\/\//i, '');
  }

  /**
   * this method will parse connection string into connection parameters
   *
   * @param {string} connection in the uri scheme (cassandra://localhost:9042)
   */
  static getConnectionParameters(connection: string) {
    const dialect = getDialectType(connection);

    if (dialect) {
      const connectionStringParser = new ConnectionStringParser({
        scheme: dialect,
        hosts: [],
      });
      let res = connectionStringParser.parse(connection);

      if (!res || Object.keys(res).length === 0) {
        try {
          // here we attempt to encode and retry parser
          let connectionParts = connection.replace(`${dialect}://`, '').split(/[:@]/);
          if (connectionParts.length === 4) {
            // there are 4 parts: username, password, host, port
            const [username, password, host, port] = connectionParts.map(encodeURIComponent);
            res = connectionStringParser.parse(
              `${dialect}://${username}:${password}@${host}:${port}`,
            );
          }
        } catch (err) {}
      }

      if (Object.keys(res).length > 0) {
        return res;
      }
    }

    // not supported
    return undefined;
  }

  static resolveTypes(
    inputItem: any,
    incomingTypeConverter?: (type: string, value: any) => string,
  ) {
    const stack: {
      item: any;
      path: string[];
      key: string;
    }[] = [{ item: inputItem, path: [], key: '' }];

    const onTypeConverter = (type: string, value: any) => {
      // override the number to be more specific (integer vs float)
      if (type === 'number') {
        if (Number.isInteger(value)) {
          type = 'integer';
        } else {
          type = 'float';
        }
      }

      if (incomingTypeConverter) {
        return incomingTypeConverter(type, value);
      }

      return type;
    };

    const columnsMap: Record<string, SqluiCore.ColumnMetaData> = {};
    const visited = new Set<string>();
    while (stack.length > 0) {
      //@ts-ignore
      const { item, path, key } = stack.pop();
      const type = Array.isArray(item) ? 'array' : typeof item;
      if (item === null || item === undefined) {
        // TODO: if item has a null or undefined, let's set the allow null flag
      } else if (type === 'object') {
        for (const targetKey of Object.keys(item)) {
          const newPath = [...path, targetKey];
          const newKey = newPath.join('/');

          if (!visited.has(newKey)) {
            stack.push({
              item: item[targetKey],
              path: newPath,
              key: newKey,
            });

            visited.add(newKey);
          }
        }
      } else {
        // TODO: figure out the max length
        columnsMap[key] = {
          ...(columnsMap[key] || {}),
          name: key,
          propertyPath: path,
          type: onTypeConverter(type, item),
          nested: path.length > 1, // whether or not this is a complex type and nested inside another JSON
        };
      }
    }

    return columnsMap;
  }

  static inferTypesFromItems(items: any[]): SqluiCore.ColumnMetaData[] {
    let columnsMap: Record<string, SqluiCore.ColumnMetaData> = {};

    for (const item of items) {
      columnsMap = {
        ...columnsMap,
        ...BaseDataAdapter.resolveTypes(item),
      };
    }

    return Object.values(columnsMap);
  }
  static inferSqlTypeFromItems(items: any[], toDialectHint?: string): SqluiCore.ColumnMetaData[] {
    let columnsMap: Record<string, SqluiCore.ColumnMetaData> = {};

    for (const item of items) {
      columnsMap = {
        ...columnsMap,
        ...BaseDataAdapter.resolveTypes(item, (type: string, val: any) => {
          switch (type) {
            case 'float':
              switch (toDialectHint) {
                case 'mysql':
                case 'mariadb':
                case 'mssql':
                case 'postgres':
                case 'postgresql':
                case 'sqlite':
                case 'cassandra':
                  return 'FLOAT';
                case 'mongodb':
                case 'redis':
                case 'cosmosdb':
                case 'aztable':
                default:
                  return type;
              }
            case 'integer':
              switch (toDialectHint) {
                case 'mysql':
                case 'mariadb':
                case 'mssql':
                case 'postgres':
                case 'postgresql':
                case 'sqlite':
                  return 'INTEGER';
                case 'cassandra':
                  return 'INT';
                case 'mongodb':
                case 'redis':
                case 'cosmosdb':
                case 'aztable':
                default:
                  return type;
              }
            case 'boolean':
              switch (toDialectHint) {
                case 'mssql':
                  return 'BIT';
                case 'mysql':
                case 'mariadb':
                case 'postgres':
                case 'postgresql':
                case 'sqlite':
                case 'cassandra':
                  return 'BOOLEAN';
                case 'mongodb':
                case 'redis':
                case 'cosmosdb':
                case 'aztable':
                default:
                  return type;
              }
            default:
              // TODO: should pick up nvarchar vs text instead of default to text
              switch (toDialectHint) {
                case 'mysql':
                case 'mariadb':
                case 'mssql':
                case 'postgres':
                case 'postgresql':
                case 'sqlite':
                case 'cassandra':
                  return 'TEXT';
                case 'mongodb':
                case 'redis':
                case 'cosmosdb':
                case 'aztable':
                default:
                  return type;
              }
          }
        }),
      };
    }

    return Object.values(columnsMap);
  }
}
