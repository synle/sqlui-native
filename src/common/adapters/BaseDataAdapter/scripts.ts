import { getDialectType } from 'src/common/adapters/DataScriptFactory';
import IDataScript from 'src/common/adapters/IDataScript';
import { SqlAction, SqluiCore } from 'typings';

export function getDivider(): SqlAction.Output {
  return {
    label: 'divider',
    skipGuide: true,
  };
}

export default abstract class BaseDataScript implements IDataScript {
  dialects: string[] = [];

  isDialectSupported(dialect?: string) {
    return !!dialect && this.dialects.indexOf(dialect) >= 0;
  }

  getConnectionFormInputs() {
    return [
      ['username', 'Username'],
      ['password', 'Password'],
      ['host', 'Host'],
      ['port', 'Port', 'optional'],
    ];
  }

  getIsTableIdRequiredForQuery() {
    return false;
  }

  getSyntaxMode() {
    return 'sql';
  }

  supportMigration() {
    return false;
  }

  supportCreateRecordForm() {
    return false;
  }

  supportEditRecordForm() {
    return false;
  }

  // dialect definitions
  getDialectType(dialect?: SqluiCore.Dialect) {
    // attempt to return the first item in the dialects / schemes
    if (dialect && this.dialects.indexOf(dialect) >= 0) {
      return dialect as SqluiCore.Dialect;
    }

    return undefined;
  }

  getDialectName(dialect?: SqluiCore.Dialect): string {
    // capitalize the first letter
    return (dialect || '').replace(/^\w/, (c) => c.toUpperCase());
  }

  getDialectIcon(dialect?: SqluiCore.Dialect): string {
    return `${process.env.PUBLIC_URL}/assets/${dialect}.png`;
  }

  // sample data script
  getTableScripts(): SqlAction.TableActionScriptGenerator[] {
    return [];
  }

  getDatabaseScripts(): SqlAction.DatabaseActionScriptGenerator[] {
    return [];
  }

  getConnectionScripts(): SqlAction.ConnectionActionScriptGenerator[] {
    return [];
  }

  getSampleConnectionString(dialect?: SqluiCore.Dialect) {
    return '';
  }

  getSampleSelectQuery(actionInput: SqlAction.TableInput): SqlAction.Output | undefined {
    return undefined;
  }

  // sample code snippet
  getCodeSnippet(
    connection: SqluiCore.ConnectionProps,
    query: SqluiCore.ConnectionQuery,
    language: SqluiCore.LanguageMode,
  ) {
    switch (language) {
      case 'javascript':
        // TODO: implement me
        return '';
      case 'python':
        // TODO: implement me
        return '';
      case 'java':
        // TODO: implement me
        return '';
      default:
        return '';
    }
  }
}
