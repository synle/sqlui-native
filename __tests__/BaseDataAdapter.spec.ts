import BaseDataAdapter from 'commons/adapters/BaseDataAdapter';

describe('BaseDataAdapter', () => {
  describe('getConnectionParameters', () => {
    test('basic input should work ', async () => {
      const config = BaseDataAdapter.getConnectionParameters('cassandra://localhost:9042');
      expect(config?.scheme).toBe('cassandra');
      expect(config?.hosts).toMatchSnapshot();
    });

    test('input with keyspace', async () => {
      const config = BaseDataAdapter.getConnectionParameters(
        'cassandra://localhost:9042/system_schema',
      );
      expect(config?.scheme).toBe('cassandra');
      expect(config?.endpoint).toBe('system_schema');
      expect(config?.hosts).toMatchSnapshot();
    });

    test('input with username and password', async () => {
      const config = BaseDataAdapter.getConnectionParameters(
        'cassandra://username:password@localhost:9042',
      );
      expect(config?.scheme).toBe('cassandra');
      expect(config?.username).toBe('username');
      expect(config?.password).toBe('password');
      expect(config?.hosts).toMatchSnapshot();
    });

    test('input with username and password and database', async () => {
      const config = BaseDataAdapter.getConnectionParameters(
        'cassandra://username:password@localhost:9042/system_schema',
      );
      expect(config?.scheme).toBe('cassandra');
      expect(config?.username).toBe('username');
      expect(config?.password).toBe('password');
      expect(config?.endpoint).toBe('system_schema');
      expect(config?.hosts).toMatchSnapshot();
    });
  });

  describe('resolveTypes', () => {
    test('primitive types only', async () => {
      const actual = BaseDataAdapter.resolveTypes({
        _id: '620d36396027dda455d46763',
        name: 'sy le',
        zipcode: 95037,
      });
      expect(actual).toStrictEqual({
        _id: { name: '_id', type: 'string' },
        name: { name: 'name', type: 'string' },
        zipcode: { name: 'zipcode', type: 'number' },
      });
    });

    test('arrays', async () => {
      const actual = BaseDataAdapter.resolveTypes({ genre: ['aa', 'bb', 'cc'] });
      expect(actual).toStrictEqual({ genre: { name: 'genre', type: 'array' } });
    });

    test('nested objects', async () => {
      const actual = BaseDataAdapter.resolveTypes({
        _id: '620d36396027dda455d46763',
        name: 'sy le',
        location: { zip: 95037, county: { name: 'santa clara' } },
        genre: ['aa', 'bb', 'cc'],
      });
      expect(actual).toStrictEqual({
        _id: { name: '_id', type: 'string' },
        genre: { name: 'genre', type: 'array' },
        'location.county.name': { name: 'location.county.name', type: 'string' },
        'location.zip': { name: 'location.zip', type: 'number' },
        name: { name: 'name', type: 'string' },
      });
    });
  });
});
