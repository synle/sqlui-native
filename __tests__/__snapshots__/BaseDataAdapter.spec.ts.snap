// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseDataAdapter getConnectionParameters basic input should work  1`] = `
Array [
  Object {
    "host": "localhost",
    "port": 9042,
  },
]
`;

exports[`BaseDataAdapter getConnectionParameters input with keyspace 1`] = `
Array [
  Object {
    "host": "localhost",
    "port": 9042,
  },
]
`;

exports[`BaseDataAdapter getConnectionParameters input with username and password 1`] = `
Array [
  Object {
    "host": "localhost",
    "port": 9042,
  },
]
`;

exports[`BaseDataAdapter getConnectionParameters input with username and password and database 1`] = `
Array [
  Object {
    "host": "localhost",
    "port": 9042,
  },
]
`;
